File: backend/internal/bff/grpc/converters/policy.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

package converters

import (
	identity_service_sdk_go "github.com/agntcy/identity-service/api/server/agntcy/identity/service/v1alpha1"
	policytypes "github.com/agntcy/identity-service/internal/core/policy/types"
	"github.com/agntcy/identity-service/internal/pkg/convertutil"
	"github.com/agntcy/identity-service/internal/pkg/ptrutil"
)

func FromPolicy(src *policytypes.Policy) *identity_service_sdk_go.Policy {
	if src == nil {
		return nil
	}

	return &identity_service_sdk_go.Policy{
		Id:          ptrutil.Ptr(src.ID),
		Name:        ptrutil.Ptr(src.Name),
		Description: ptrutil.Ptr(src.Description),
		AssignedTo:  ptrutil.Ptr(src.AssignedTo),
		Rules:       convertutil.ConvertSlice(src.Rules, FromRule),
		CreatedAt:   newTimestamp(&src.CreatedAt),
	}
}

func FromRule(src *policytypes.Rule) *identity_service_sdk_go.Rule {
	if src == nil {
		return nil
	}

	return &identity_service_sdk_go.Rule{
		Id:            ptrutil.Ptr(src.ID),
		Name:          ptrutil.Ptr(src.Name),
		Description:   ptrutil.Ptr(src.Description),
		NeedsApproval: ptrutil.Ptr(src.NeedsApproval),
		Tasks:         convertutil.ConvertSlice(src.Tasks, FromTask),
		Action:        ptrutil.Ptr(identity_service_sdk_go.RuleAction(src.Action)),
		CreatedAt:     newTimestamp(&src.CreatedAt),
	}
}

func FromTask(src *policytypes.Task) *identity_service_sdk_go.Task {
	if src == nil {
		return nil
	}

	return &identity_service_sdk_go.Task{
		Id:       ptrutil.Ptr(src.ID),
		Name:     ptrutil.Ptr(src.Name),
		AppId:    ptrutil.Ptr(src.AppID),
		ToolName: ptrutil.Ptr(src.ToolName),
	}
}

File: frontend/src/types/api/policy.ts
/**
 * Copyright 2025 Cisco Systems, Inc. and its affiliates
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  V1Alpha1ListPoliciesResponse,
  V1Alpha1ListRulesResponse,
  V1Alpha1PagedResponse,
  V1Alpha1Policy,
  V1Alpha1Rule,
  V1Alpha1Task,
  V1Alpha1CreatePolicyRequest,
  PolicyServiceCreateRuleBody,
  V1Alpha1RuleAction
} from '@/api/generated/identity/policy_service.swagger.api';

export type {
  V1Alpha1ListPoliciesResponse as ListPoliciesResponse,
  V1Alpha1ListRulesResponse as ListRulesResponse,
  V1Alpha1PagedResponse as PagedResponse,
  V1Alpha1Policy as Policy,
  V1Alpha1Rule as Rule,
  V1Alpha1Task as Task,
  V1Alpha1CreatePolicyRequest as CreatePolicyRequest,
  PolicyServiceCreateRuleBody as CreateRuleBody
};

export {V1Alpha1RuleAction as RuleAction};
